### PR TITLE
fix: Overflow checks for constant folding

### LINKED ISSUE #3395: Unsigned integers do not overflow when they are defined as constants
### Aim

Define constant unsigned integers, and summing them beyond their type boundaries.

### Expected Behavior

E.g. given `u8`'s boundary goes from 0 to 255, both of the following examples should fail:

1. Adding `u8` inputs together to beyond the type boundary:

```rust
fn main(x : u8, y : u8) {
    x + y;
}
```

with

```toml
x = "255"
y = "255"
```

2. Adding `u8` constants together to beyond the type boundary:

```rust
fn main() {
    let x : u8 = 255;
    let y : u8 = 255;
    x + y;
}
```

### Bug

Proving (1) fails with `Assertion failed: 'attempt to add with overflow'` as expected.

Proving (2) however did not result in any errors.

### Installation Method

Binary

### Nargo Version

(Nightly)

nargo version = 0.18.0 noirc version = 0.18.0+a0985412594c5cbd35551afc7e94cb444723adb0 (git version hash: a0985412594c5cbd35551afc7e94cb444723adb0, is dirty: false)

### Additional Context

Discovered while drafting examples in https://github.com/noir-lang/noir/pull/3393.

### Would you like to submit a PR for this Issue?

No

### Support Needs

_No response_

### PR DESCRIPTION
**Pull request by guipublic**
# Description

## Problem\*

Resolves #3395

## Summary\*

Check for overflow during constant folding

## Documentation\*

Check one:
- [X] No documentation needed.
- [ ] Documentation included in this PR.
- [ ] **[Exceptional Case]** Documentation to be submitted in a separate PR.

# PR Checklist\*

- [X] I have tested the changes locally.
- [X] I have formatted the changes with [Prettier](https://prettier.io/) and/or `cargo fmt` on default settings.

### CODE REVIEW

File: compiler/noirc_evaluator/src/ssa/ir/instruction.rs [hunk 1]
```diff
@@ -847,9 +847,16 @@ fn eval_constant_binary_op(
             if matches!(operator, BinaryOp::Div | BinaryOp::Mod) && rhs == 0 {
                 return None;
             }
-
             let result = function(lhs, rhs);
```
Comment (TomAFrench) [inline compiler/noirc_evaluator/src/ssa/ir/instruction.rs:850]:
```suggestion
            let result = function(lhs, rhs)?;
```

We can `?` here to return `None` values

File: compiler/noirc_evaluator/src/ssa/ir/instruction.rs [hunk 2]
```diff
@@ -847,9 +847,16 @@ fn eval_constant_binary_op(
...
-            truncate(result, *bit_size).into()
+            if let Some(result) = result {
+                // Check for overflow
+                if result >= 2u128.pow(*bit_size) {
+                    return None;
+                }
+                result.into()
+            } else {
+                return None;
+            }
```
Comment (TomAFrench) [inline compiler/noirc_evaluator/src/ssa/ir/instruction.rs:859]:
```suggestion
            // Check for overflow
            if result >= 2u128.pow(*bit_size) {
                return None;
            }
            result.into()
```

which allows us to remove a layer of nesting here.

File: compiler/noirc_evaluator/src/ssa/ir/instruction.rs [hunk 3]
```diff
@@ -870,9 +877,20 @@ fn eval_constant_binary_op(
             }
 
             let result = function(lhs, rhs);
```
Comment (TomAFrench) [inline compiler/noirc_evaluator/src/ssa/ir/instruction.rs:879]: Can do the same here.

**Review by TomAFrench — APPROVED** [review]
Looks good, we can remove some nesting however.

File: compiler/noirc_evaluator/src/ssa/ir/instruction.rs (hunk 2 above)
Comment (guipublic) [reply compiler/noirc_evaluator/src/ssa/ir/instruction.rs]: Done

File: compiler/noirc_evaluator/src/ssa/ir/instruction.rs (hunk 3 above)
Comment (guipublic) [reply compiler/noirc_evaluator/src/ssa/ir/instruction.rs]: Done

**Pull request merged**